Move volume_expand tests to storage e2e testsuites

mkimuram opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR move volume_expand tests to storage e2e testsuites. It will allow sharing volume_expand tests with csi-drivers and other in-tree drivers than aws and gce, as a result, test coverage will increase.

**Which issue(s) this PR fixes**:
Fixes #77999

**Special notes for your reviewer**:
/sig storage
cc @gnufied @msau42

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

mkimuram committed:
Move volume_expand tests to storage e2e testsuites

mkimuram commented:
@gnufied 

Could you review this refactoring?

Below four tests are executed in CI, by the codes.
```
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block volmode)] volume-expand should not allow expansion of pvcs without AllowVolumeExpansion property
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] volume-expand Verify if editing PVC allows resize
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)] volume-expand should not allow expansion of pvcs without AllowVolumeExpansion property
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)(allowExpansion)] volume-expand Verify if editing PVC allows resize
```

Note that:
- Filesystem and block volume tests now shares the same test logic, except condition check.
- Coverage for in-tree driver tests are not increased for CI. (Dynamic provision tests are only tested for NFS driver, but it doesn't seem to support volume expansion. )
- Coverage for csi driver tests are not increased for CI for now. (We need to wait until gce csi driver to support volume expansion. But other csi drivers might benefit by using this framework from out-of-tree.)

mkimuram commented:
@gnufied 

Rebased. This should now run below 6 test cases. PTAL

```
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block volmode)] volume-expand should not allow expansion of pvcs without AllowVolumeExpansion property
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] Verify if offline PVC expansion works
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] should resize volume when PVC is edited while pod is using it
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)] volume-expand should not allow expansion of pvcs without AllowVolumeExpansion property
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)(allowExpansion)] Verify if offline PVC expansion works
[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)(allowExpansion)] should resize volume when PVC is edited while pod is using it
```

gnufied commented:
Taking a look.

/assign

gnufied commented on test/e2e/storage/testsuites/volume_expand.go:
```diff
@@ -0,0 +1,344 @@
+const (
+	resizePollInterval = 2 * time.Second
+	// total time to wait for cloudprovider or file system resize to finish
+	totalResizeWaitPeriod = 20 * time.Minute
```
Now that we are moving this here, this can be reduced to something like 10 minutes. These defaults were too high because we had incorrect assumption about when EBS disks are "finished" resizing.

mkimuram replied:
Updated.

gnufied commented on test/e2e/storage/flexvolume_mounted_volume_resize.go:
```diff
@@ -35,6 +36,11 @@ import (
+const (
+	// total time to wait for cloudprovider or file system resize to finish
+	totalResizeWaitPeriod = 20 * time.Minute
+)
```
Lets reduce this time to like 5 minutes.

mkimuram replied:
Updated.

gnufied commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -1148,6 +1148,7 @@ func InitGcePdDriver() testsuites.TestDriver {
 				testsuites.CapFsGroup:     true,
 				testsuites.CapBlock:       true,
 				testsuites.CapExec:        true,
+				testsuites.CapExpansion:   true,
```
We can perhaps fix this in a follow up but there would be volumes that only require control plane expansion and volume types that require both and hence this capability does not expresses that intent very well.

For example - if we were to enable these tests for glusterfs driver, there won't be a file system resize happening and tests may need some tweaking.

mkimuram replied:
I'm implementing controller side code of csi-driver-nfs [here](https://github.com/kubernetes-csi/csi-driver-nfs/pull/13), and this driver also won't require controller expansion implementation.

In that case, could we just set true for such driver's capability for the test code, or do we need to add special handling for it?

gnufied replied:
I think this is fine for now. Once we start adding tests for driver which need to skip control or node expansion, we can revisit this.

msau42 replied:
Now that we run the testsuite outside of kubernetes, these capabilities are becoming more of an external API, and we should not try to change them too much.  Can we pre add NodeExpansion and ControllerExpansion capabilities now to anticipate the future tests?

mkimuram replied:
Separated `NodeExpansion` and `ControllerExpansion` capabilities. I guess that only drivers that have `NodeExpansion` should transit their PVC conditions, so I changed the test code as so. Please check if it is correct.

gnufied commented:
@maazkhan there appears to be a verify failure:

```
Expect(err).To(HaveOccurred()) or gomega.Expect(err).To(gomega.HaveOccurred())
```
Can you fix that please?

mkimuram commented:
@gnufied 

Updated the commit. PTAL

> Expect(err).To(HaveOccurred()) or gomega.Expect(err).To(gomega.HaveOccurred())

For above comment, I've replaced  them with framework.ExpectNoError and framework.ExpectError for these particular e2e tests, does it fix what you meant?

(Also, I made another PR to https://github.com/kubernetes/kubernetes/pull/78916 fix it for entire e2e tests.)

gnufied commented:
/lgtm

msau42 commented on test/e2e/storage/flexvolume_mounted_volume_resize.go:
```diff
@@ -178,10 +184,18 @@ var _ = utils.SIGDescribe("Mounted flexvolume expand[Slow]", func() {
 	})
 })
+
+func createStorageClass(t testsuites.StorageClassTest, ns string, suffix string, c clientset.Interface) (*storage.StorageClass, error) {
```
We create storageclasses for many tests. Is there a common function we can use?

msau42 commented on test/e2e/storage/testpatterns/testpattern.go:
```diff
@@ -55,12 +56,14 @@ var (
+	FsType         string                      // Fstype of the volume
+	VolMode        v1.PersistentVolumeMode     // PersistentVolumeMode of the volume
+	SnapshotType   TestSnapshotType            // Snapshot type of the snapshot
+	BindingMode    storagev1.VolumeBindingMode // VolumeBindingMode of the volume
```
Is this used anywhere?

msau42 replied:
Let's open up a separate issue to discuss this further but I want to move away from being able to toggle/customize StorageClass arguments in the framework tests.  Now that these tests are also used externally by CSI driver developers, I see a use case where a vendor wants to test their plugin with specific recommended/default storage classes that they provide.

And in some cases, toggling these parameters automatically in the test suite may not work for the storage plugin.  For example, immediate binding mode is not going to work in a multi-zone cluster.  It would be better in this case if the vendor can completely control all the parameters instead of the test suite.

mkimuram commented:
@msau42 

Thank you for your review. Fixed. PTAL

davidz627 commented:
What is the status on this - are we waiting on another round of reviews from anyone or are we pending on rebase?

mkimuram commented:
@davidz627 

Thank you for taking care of this PR.

> What is the status on this - are we waiting on another round of reviews from anyone or are we pending on rebase?

I expect final review from @msau42  and @gnufied , especially for below two commits:
 - "Separate CapControllerExpansion and CapNodeExpansion"
 - "Avoid adding another way to create StorageClass"

If they are good, I will squash the commits, then this PR should be ready to be merged.

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -56,6 +56,7 @@ var testSuites = []func() testsuites.TestSuite{
 	testsuites.InitSubPathTestSuite,
 	testsuites.InitProvisioningTestSuite,
 	testsuites.InitMultiVolumeTestSuite,
+	testsuites.InitVolumeExpandTestSuite,
```
Can you add this here too: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/external/external.go#L48

msau42 commented:
/approve

mkimuram commented:
@msau42 

Thank you for comment and approval.
I've added `VolumeExpandTestSuite` to `e2e/storage/external/external.go` and squashed the commit.

mkimuram commented:
/retest

msau42 commented:
Will let @gnufied do final review. Are these covering offline resize, or is online resize also here?  Can you open a tracking issue if we're missing some coverage?

mkimuram commented:
@msau42 @gnufied 

> Can you open a tracking issue if we're missing some coverage?

I've opened it as #79445 . Please feel free to edit it.

gnufied commented:
/lgtm

davidz627 commented:
/retest
